core/server/common/src/main/java/alluxio/master/journal/raft/JournalStateMachine.java

```diff
@@ -255,42 +260,64 @@ protected void afterServerClosing() {
     }
   }
 
+  /**
+   * Allows leader to take snapshots. This is used exclusively for the
+   * `bin/alluxio fsadmin journal checkpoint` command.
+   * @param manager allows the state machine to take a snapshot as leader by using it
+   */
+  void allowLeaderSnapshots(StateLockManager manager) {
+    mStateLockManagerRef.set(manager);
+  }
+
+  void disallowLeaderSnapshots() {
+    mStateLockManagerRef.set(null);
+  }
+
   @Override
   public long takeSnapshot() {
-    if (mIsLeader) {
-      SAMPLING_LOG.info("Calling take snapshot on leader");
-      try {
-        RaftGroup group;
-        try (LockResource ignored = new LockResource(mGroupLock)) {
-          if (mServerClosing) {
-            return RaftLog.INVALID_LOG_INDEX;
-          }
-          // These calls are protected by mGroupLock and mServerClosing
-          // as they will access the lock in RaftServerProxy.java
-          // which is also accessed during raft server shutdown which
-          // can cause a deadlock as the shutdown takes the lock while
-          // waiting for this thread to finish
-          Preconditions.checkState(mServer.getGroups().iterator().hasNext());
-          group = mServer.getGroups().iterator().next();
-        }
-        Preconditions.checkState(group.getGroupId().equals(mRaftGroupId));
-        if (group.getPeers().size() < 2) {
-          SAMPLING_LOG.warn("No follower to perform delegated snapshot. Please add more masters to "
-              + "the quorum or manually take snapshot using 'alluxio fsadmin journal checkpoint'");
+    long index;
+    StateLockManager stateLockManager = mStateLockManagerRef.get();
+    if (!mIsLeader) {
+      index = takeLocalSnapshot();
+    } else if (stateLockManager != null) {
+      // the leader has been allowed to take a local snapshot by being given a non-null
+      // StateLockManager through the #allowLeaderSnapshots method
+      try (LockResource stateLock = stateLockManager.lockExclusive(StateLockOptions.defaults())) {
+        index = takeLocalSnapshot();
```

Comment:
I think inside the call to takeLocalSnapshot you may have to reverify you are in the correct state since the checks are made outside of the synchronized block.